File: src/commonMain/kotlin/io/kontainers/model/ContainerStats.kt
package io.kontainers.model

import kotlinx.serialization.Serializable
import kotlinx.serialization.Transient

/**
 * Represents container resource usage statistics.
 */
@Serializable
data class ContainerStats(
    val containerId: String,
    val timestamp: Long,
    val cpuUsage: Double,
    val memoryUsage: Long,
    val memoryLimit: Long,
    val networkRx: Long,
    val networkTx: Long,
    val blockRead: Long,
    val blockWrite: Long
) {
    /**
     * Calculates memory usage as a percentage of the limit.
     * Returns 0.0 if there is no memory limit.
     */
    @Transient
    val memoryUsagePercentage: Double
        get() = if (memoryLimit > 0) memoryUsage.toDouble() / memoryLimit.toDouble() else 0.0
}

File: shared/src/models/container-stats.ts
/**
 * Container statistics models for Kontainers application
 */

import { HealthStatus } from './health';

/**
 * Represents container resource usage statistics.
 */
export interface ContainerStats {
  containerId: string;
  timestamp: number;
  cpuUsage: number;
  memoryUsage: number;
  memoryLimit: number;
  networkRx: number;
  networkTx: number;
  blockRead: number;
  blockWrite: number;
  
  // Calculated properties
  memoryUsagePercentage?: number;
}

/**
 * Represents detailed container resource usage statistics.
 * Extends the basic ContainerStats with additional metrics for enhanced monitoring.
 */
export interface DetailedContainerStats extends Pick<ContainerStats, 
  'containerId' | 'timestamp' | 'cpuUsage' | 'memoryUsage' | 'memoryLimit' | 
  'networkRx' | 'networkTx' | 'blockRead' | 'blockWrite'> {
  
  // Enhanced metrics
  cpuKernelUsage: number;
  cpuUserUsage: number;
  cpuSystemUsage: number;
  cpuOnlineCpus: number;
  memoryCache: number;
  memorySwap: number;
  memorySwapLimit: number;
  networkPacketsRx: number;
  networkPacketsTx: number;
  networkDroppedRx: number;
  networkDroppedTx: number;
  networkErrorsRx: number;
  networkErrorsTx: number;
  blockReadOps: number;
  blockWriteOps: number;
  pids: number;
  restartCount: number;
  healthStatus: HealthStatus;
  healthCheckLogs: string[];
  
  // Calculated properties
  memoryUsagePercentage?: number;
  swapUsagePercentage?: number;
  networkThroughput?: number;
  blockThroughput?: number;
}